File: lib/src/main/java/org/simplepresenter/PresenterFragment.java
package org.simplepresenter;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;

public abstract class PresenterFragment<P extends Presenter> extends Fragment implements PresenterView<P> {

    private final PresenterViewDelegate<P> delegate = new PresenterViewDelegate<>(this);

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        delegate.onCreate(savedInstanceState);
    }

    @Override
    public void onPause() {
        super.onPause();
        delegate.onPause();
    }

    @Override
    public void onResume() {
        super.onResume();
        delegate.onResume();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        delegate.onSaveInstanceState(outState);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        delegate.onDestroy();
    }

    @NonNull
    @Override
    public P getPresenter() {
        if (delegate.getPresenter() == null) {
            throw new IllegalStateException();
        }
        return delegate.getPresenter();
    }
}

File: lib/src/main/java/org/simplepresenter/PresenterFragment.kt
package org.simplepresenter

import android.os.Bundle
import android.support.v4.app.Fragment

abstract class PresenterFragment<P : Presenter> : Fragment(), PresenterView<P> {

    val delegate = PresenterViewDelegate<P>(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        delegate.onCreate(savedInstanceState)
    }

    override fun onPause() {
        super.onPause()
        delegate.onPause()
    }

    override fun onResume() {
        super.onResume()
        delegate.onResume()
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        delegate.onSaveInstanceState(outState)
    }

    override fun onDestroy() {
        super.onDestroy()
        delegate.onDestroy()
    }

    override val presenter: P
        get() = delegate.presenter ?: throw IllegalStateException()
}